### PR TITLE
Conversion modulo eta

### PR DESCRIPTION
**Pull request by 01mf02**
To quote Makarius Wenzel:

> Isabelle proof terms are always understood modulo beta-eta.

This commit adds an option to Lambdapi allowing typechecking modulo eta, thus allowing checking of Isabelle proof terms.

### CODE REVIEW

**Comment by rlepigre** [comment]
Looks great to me. However it would be nice to fix the test since it is currently not run with the new option. I propose adding a folder `tests/OK+eta-conv` in which all files should expect the `--eta-conv` option. If you do that, then you'll need to modify `GNUmakefile` accordingly, following what is already done.

File: src/eval.ml [hunk 1]
```diff
@@ -128,6 +131,10 @@ and eq_modulo : term -> term -> bool = fun a b ->
     | (Abst(a1,b1), Abst(a2,b2)) ->
         let (_,b1,b2) = Bindlib.unbind2 b1 b2 in
         eq_modulo ((a1,a2)::(b1,b2)::l)
+    | (Abst(_ ,b ), t          )
+    | (t          , Abst(_ ,b )) when Pervasives.(! eta_conv) ->
+        let (x,b) = Bindlib.unbind b in
+        eq_modulo ((b,(Appl(t, Vari(x))))::l)
```
Comment (fblanqui) [inline src/eval.ml]: The number of parentheses can be minimized.

**Comment by fblanqui** [comment]
Instead of using command line options, wouldn't it better to add some command in a file to tell LP to use eta or not? Like `set flag "eta" on/off`.

**Comment by rlepigre** [comment]
> Instead of using command line options, wouldn't it better to add some command in a file to tell LP to use eta or not? Like `set flag "eta" on/off`.

I'm not sure this is a good idea, because this would allow mixing code that relies on eta and code that does not. But we can of course do that. Since this is still a quite experimental feature, I would vote for keeping things as they are for now.

**Comment by fblanqui** [comment]
What's the problem with mixing code using eta and code not using eta? Note that eta could be seen as a rewriting rule if we allowed rules headed by abstractions: `rule λx,M x → M`.

**Comment by 01mf02** [comment]
Following Frédéric's suggestion, I made equality modulo eta a flag instead of a command-line option. This follows the intuition that this option should always be enabled for files that require it, and it solves the issue of the test failing.
I named the flag `eta_equality`, following [the equivalent flag in Agda](https://github.com/agda/agda/issues/2436).

**Comment by 01mf02** [comment]
Sorry, forgot to integrate Frédéric's suggestion to reduce the number of parentheses.

**Comment by rlepigre** [comment]
No problem, thanks @01mf02!

**Pull request merged**